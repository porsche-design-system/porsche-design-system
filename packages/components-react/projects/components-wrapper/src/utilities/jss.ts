export * from '@porsche-design-system/components-js/utilities/jss';
