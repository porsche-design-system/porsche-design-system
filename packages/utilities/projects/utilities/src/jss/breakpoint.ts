export type Breakpoint = 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl';

export const breakpoint: { [key in Breakpoint]: number } = {
  xxs: 0, // TODO: might be base to be in sync with breakpoint customizable
  xs: 480,
  s: 760,
  m: 1000,
  l: 1300,
  xl: 1760,
  xxl: 1920, // TODO: xxl missing in breakpoint customizable
};
