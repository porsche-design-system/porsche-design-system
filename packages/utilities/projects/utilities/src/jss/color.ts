export type ThemeDefault = 'light' | 'dark';
export type ThemeElectric = 'light-electric' | 'dark-electric';
export type Theme = ThemeDefault | ThemeElectric;

type ColorTheme = {
  brand: string;
  base: string;
  background: {
    base: string;
    surface: string;
    shading: string;
  };
  contrast: {
    high: string;
    medium: string;
    low: string;
  };
  notification: {
    success: string;
    successSoft: string;
    warning: string;
    warningSoft: string;
    error: string;
    errorSoft: string;
    neutral: string;
    neutralSoft: string;
  };
  state: {
    hover: string;
    active: string;
    focus: string;
    disabled: string;
  };
};

type ColorExternalKey =
  | 'facebook'
  | 'google'
  | 'instagram'
  | 'kakaotalk'
  | 'linkedin'
  | 'naver'
  | 'pinterest'
  | 'reddit'
  | 'tiktok'
  | 'twitter'
  | 'wechat'
  | 'whatsapp'
  | 'xing'
  | 'youtube';
type ColorExternal = { [key in ColorExternalKey]: string };

export const themeLight: ColorTheme = {
  brand: '#d5001c',
  base: '#000',
  background: {
    base: '#fff',
    surface: '#f2f2f2',
    shading: 'rgba(14, 20, 24, 0.9)',
  },
  contrast: {
    high: '#323639',
    medium: '#626669',
    low: '#e3e4e5',
  },
  notification: {
    success: '#018a16',
    successSoft: '#e5f3e7',
    warning: '#ff9b00',
    warningSoft: '#fff5e5',
    error: '#e00000',
    errorSoft: '#fae6e6',
    neutral: '#0061bd',
    neutralSoft: '#e5eff8',
  },
  state: {
    hover: '#d5001c',
    active: '#d5001c',
    focus: 'currentColor',
    disabled: '#96989a',
  },
};

export const themeDark: ColorTheme = {
  brand: '#d5001c',
  base: '#fff',
  background: {
    base: '#0e1418',
    surface: '#262b2e',
    shading: 'rgba(14, 20, 24, 0.9)',
  },
  // TODO: remove neutral?
  contrast: {
    high: '#e3e4e5',
    medium: '#b0b1b2',
    low: '#4a4e51',
  },
  notification: {
    success: '#01ba1d',
    successSoft: '#bfeec6',
    warning: '#ff9b00',
    warningSoft: '#ffe6bf',
    error: '#fc1717',
    errorSoft: '#fec5c5',
    neutral: '#2193ff',
    neutralSoft: '#c7e4ff',
  },
  state: {
    hover: '#ff0223',
    active: '#ff0223',
    focus: 'currentColor',
    disabled: '#7c7f81',
  },
};

export const themeLightElectric: ColorTheme = {
  ...themeLight,
  brand: '#00b0f4',
  state: {
    ...themeLight.state,
    hover: '#00b0f4',
    active: '#00b0f4',
  },
};

export const themeDarkElectric: ColorTheme = {
  ...themeDark,
  brand: '#00b0f4',
  state: {
    ...themeDark.state,
    hover: '#00b0f4',
    active: '#00b0f4',
  },
};

export const colorExternal: ColorExternal = {
  facebook: '#1877f2',
  google: '#4285f4',
  instagram: '#e1306c',
  kakaotalk: '#fae300',
  linkedin: '#0077b5',
  naver: '#03cf5d',
  pinterest: '#e60023',
  reddit: '#ff4500',
  tiktok: '#fe2c55',
  twitter: '#1da1f2',
  wechat: '#1aad19',
  whatsapp: '#25d366',
  xing: '#006567',
  youtube: '#ff0000',
};

export const color = {
  light: themeLight,
  dark: themeDark,
  'light-electric': themeLightElectric,
  'dark-electric': themeDarkElectric,
  external: colorExternal,
};
