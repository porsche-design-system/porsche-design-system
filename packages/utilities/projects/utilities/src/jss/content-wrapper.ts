import type { JssStyle } from 'jss';
import { mediaQueryMin } from './media-query';

type ContentWrapperWidth = 'basic' | 'extended' | 'fluid';

const widthMap: { [key in ContentWrapperWidth]?: JssStyle } = {
  basic: {
    maxWidth: '96rem',
    padding: '0 7vw',
    [mediaQueryMin('xl')]: {
      padding: '0 10vw',
    },
    [mediaQueryMin('xxl')]: {
      padding: '0 12rem',
    },
  },
  extended: {
    maxWidth: '120rem',
  },
};

export const getContentWrapperJssStyle = (width: ContentWrapperWidth): JssStyle => {
  return {
    margin: '0 auto',
    width: '100%',
    minWidth: 0, // to handle automatic minimum size, in case it's used within flex mode
    ...widthMap[width],
  };
};
