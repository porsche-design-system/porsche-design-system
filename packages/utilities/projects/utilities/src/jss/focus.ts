import type { JssStyle } from 'jss';

type GetFocusStylesOptions = {
  color?: string;
  offset?: string;
};

export const getFocusJssStyle = (opts?: GetFocusStylesOptions): JssStyle => {
  return {
    outline: 'transparent solid 1px',
    outlineOffset: opts?.offset || '2px',
    '&::-moz-focus-inner': {
      border: '0',
    },
    '&:focus': {
      outlineColor: opts?.color || 'currentColor',
      '&:not(:focus-visible)': {
        outlineColor: 'transparent',
      },
    },
  };
};
