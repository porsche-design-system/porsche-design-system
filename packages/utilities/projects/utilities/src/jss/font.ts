export type FontSize = 'xSmall' | 'small' | 'medium' | 'large' | 'xLarge';
export type FontWeight = 'thin' | 'regular' | 'semibold' | 'bold';
export type FontSizeLineHeight = { fontSize: string; lineHeight: number };
type Font = {
  family: string;
  weight: { [key in FontWeight]: number };
  size: { [key in FontSize]: FontSizeLineHeight };
};

export const fontFamily = "'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif";
export const fontWeight: { [key in FontWeight]: number } = {
  thin: 100,
  regular: 400,
  semibold: 600,
  bold: 700,
};
export const fontSize: { [key in FontSize]: FontSizeLineHeight } = {
  xSmall: {
    fontSize: '0.75rem',
    lineHeight: 1.6666666667,
  },
  small: {
    fontSize: '1rem',
    lineHeight: 1.5,
  },
  medium: {
    fontSize: '1.5rem',
    lineHeight: 1.5,
  },
  large: {
    fontSize: '2.25rem',
    lineHeight: 1.3333333333,
  },
  xLarge: {
    fontSize: '3.25rem',
    lineHeight: 1.2307692308,
  },
};

export const font: Font = {
  family: fontFamily,
  weight: fontWeight,
  size: fontSize,
};
