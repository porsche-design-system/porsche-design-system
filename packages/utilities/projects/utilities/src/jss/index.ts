export * from './color/color';
export * from './content-wrapper';
export * from './focus';
export * from './font/font';
export * from './media-query';
export * from './screen-reader';
export * from './spacing';
export * from './typography';
