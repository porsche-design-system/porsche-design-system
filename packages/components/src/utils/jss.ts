import type { TagName } from '@porsche-design-system/shared';
import type { BreakpointCustomizable } from './breakpoint-customizable';
import type { JssStyle, Styles } from 'jss';
import { create } from 'jss';
import jssPluginCamelCase from 'jss-plugin-camel-case';
import jssPluginGlobal from 'jss-plugin-global';
import jssPluginNested from 'jss-plugin-nested';
import jssPluginSortMediaQueries from 'jss-plugin-sort-css-media-queries';
import { mediaQueryMin } from '@porsche-design-system/utilities-v2';
import type { Breakpoint } from '@porsche-design-system/utilities-v2';
import { parseJSON } from './breakpoint-customizable';
import { getShadowRootHTMLElement } from './dom';
import { addImportantToEachRule } from '../styles';
import { getTagName, getTagNameWithoutPrefix } from '.';

// NOTE: handpicked selection of plugins from jss-preset-default
const jss = create({
  plugins: [
    jssPluginGlobal(),
    jssPluginNested(),
    jssPluginCamelCase(),
    jssPluginSortMediaQueries({ combineMediaQueries: true }),
  ],
});

export const getCss = (jssStyles: Styles): string =>
  jss
    .createStyleSheet(jssStyles, {
      generateId: (rule) => rule.key,
    })
    .toString();

export const supportsConstructableStylesheets = (): boolean => {
  try {
    return typeof new CSSStyleSheet().replaceSync === 'function';
  } catch {
    return false;
  }
};

// determine it once
const hasConstructableStylesheetSupport = supportsConstructableStylesheets();
// getter for easy mocking
export const getHasConstructableStylesheetSupport = (): boolean => hasConstructableStylesheetSupport;

type CssCacheMap = Map<string, string>;
export const componentCssMap = new Map<TagName, CssCacheMap>();

export const getCachedComponentCss = <T extends (...p: any[]) => string>(
  host: HTMLElement,
  getComponentCss: T,
  ...args: Parameters<T>
): string => {
  const tagName = getTagNameWithoutPrefix(host);

  if (!componentCssMap.has(tagName)) {
    componentCssMap.set(tagName, new Map());
  }

  const id = args.map((arg) => (typeof arg === 'object' ? JSON.stringify(arg) : arg)).join('|');
  const cache = componentCssMap.get(tagName);

  if (!cache.has(id)) {
    cache.set(id, getComponentCss(...args));
  }

  return cache.get(id);
};

export const attachComponentCss = <T extends (...p: any[]) => string>(
  host: HTMLElement,
  getComponentCss: T,
  ...args: Parameters<T>
): void => {
  const css = getCachedComponentCss(host, getComponentCss, ...args);

  if (getHasConstructableStylesheetSupport()) {
    const [sheet] = host.shadowRoot.adoptedStyleSheets;
    if (sheet) {
      sheet.replaceSync(css);
    } else {
      const newSheet = new CSSStyleSheet();
      newSheet.replaceSync(css);
      host.shadowRoot.adoptedStyleSheets = [newSheet];
    }
  } else {
    // NOTE: fallback for Firefox and Safari
    // old style needs to be removed and added again in safari to be recognized
    getShadowRootHTMLElement(host, 'style[jss]')?.remove();

    const styleEl = document.createElement('style');
    styleEl.setAttribute('jss', '');
    styleEl.innerHTML = css;
    host.shadowRoot.prepend(styleEl);
  }
};

export const buildSlottedStyles = (host: HTMLElement, jssStyle: JssStyle): Styles<'@global'> => ({
  '@global': {
    [getTagName(host)]: addImportantToEachRule(jssStyle),
  },
});

export type GetJssStyleFunction = (value?: any) => JssStyle;

export const buildResponsiveStyles = <T>(
  rawValue: BreakpointCustomizable<T>,
  getJssStyle: GetJssStyleFunction
): Styles => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  const value = parseJSON(rawValue as any);

  return typeof value === 'object'
    ? Object.keys(value)
        // base styles are applied on root object, responsive styles are nested within
        // hence it is used as the initial object within reduce function
        .filter((key) => key !== 'base')
        .reduce(
          (result, breakpointValue: Breakpoint) => ({
            ...result,
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            [mediaQueryMin(breakpointValue as any)]: getJssStyle(value[breakpointValue]) as Styles,
          }),
          getJssStyle(value.base) as Styles
        )
    : (getJssStyle(value) as Styles);
};

export const isObject = <T extends Record<string, any>>(obj: T): boolean =>
  typeof obj === 'object' && !Array.isArray(obj);

// NOTE: taken from https://stackoverflow.com/a/48218209
export const mergeDeep = <T extends Record<string, any>>(...objects: T[]): T => {
  return objects.reduce((prev, obj) => {
    Object.keys(obj).forEach((key: keyof T) => {
      const pVal = prev[key];
      const oVal = obj[key];

      if (isObject(pVal) && isObject(oVal)) {
        prev[key] = mergeDeep(pVal, oVal);
      } else {
        prev[key] = oVal;
      }
    });

    return prev;
  }, {} as T);
};
