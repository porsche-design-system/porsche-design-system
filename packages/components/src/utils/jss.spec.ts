import {
  attachComponentCss,
  buildResponsiveHostStyles,
  buildResponsiveStyles,
  buildSlottedStyles,
  getCachedComponentCss,
  getCss,
  isObject,
  mergeDeep,
  supportsConstructableStylesheets,
  componentCssMap,
} from './jss';
import * as jssUtils from './jss';
import type { JssStyle, Styles } from 'jss';

describe('getCss()', () => {
  const data: { input: Styles; result: string }[] = [
    {
      input: { ':host': { display: 'block', marginLeft: '5px' } },
      result: ':host {\n  display: block;\n  margin-left: 5px;\n}',
    },
    {
      input: { ':host': { display: 'block', marginLeft: '5px !important' } },
      result: ':host {\n  display: block;\n  margin-left: 5px !important;\n}',
    },
    {
      input: { ':host': { display: 'block', width: '500px', transition: 'width .25s ease' } },
      result: ':host {\n  display: block;\n  width: 500px;\n  transition: width .25s ease;\n}',
    },
    {
      input: {
        ':host': { display: 'block', marginLeft: '5px !important' },
        '@media (min-width: 760px)': { ':host': { marginRight: '5px !important' } },
      },
      result:
        ':host {\n  display: block;\n  margin-left: 5px !important;\n}\n@media (min-width: 760px) {\n  :host {\n    margin-right: 5px !important;\n  }\n}\n',
    },
    {
      input: {
        ':host': { display: 'block', marginLeft: '5px !important' },
        '@media (min-width: 760px)': { ':host': { marginRight: '5px !important' } },
        '@media (min-width: 1000px)': { ':host': { marginRight: '10px !important' } },
      },
      result:
        ':host {\n  display: block;\n  margin-left: 5px !important;\n}\n@media (min-width: 760px) {\n  :host {\n    margin-right: 5px !important;\n  }\n}\n\n@media (min-width: 1000px) {\n  :host {\n    margin-right: 10px !important;\n  }\n}\n',
    },
    {
      input: {
        ':host': { display: 'block', marginLeft: '5px !important' },
        '@media (min-width: 1000px)': { ':host': { marginRight: '10px !important' } },
        '@media (min-width: 760px)': { ':host': { marginRight: '5px !important' } },
      },
      result:
        ':host {\n  display: block;\n  margin-left: 5px !important;\n}\n@media (min-width: 760px) {\n  :host {\n    margin-right: 5px !important;\n  }\n}\n\n@media (min-width: 1000px) {\n  :host {\n    margin-right: 10px !important;\n  }\n}\n',
    },
    {
      input: { '@global': { div: { display: 'block' } } },
      result: 'div {\n  display: block;\n}',
    },
  ];
  it.each(
    data.map(({ input, result }) => [
      JSON.stringify(input), // for test description
      JSON.stringify(result), // for test description
      input,
      result,
    ])
  )(`should transform '%s' to %s`, (_, __, input: Styles, result: string) => {
    expect(getCss(input)).toBe(result);
  });
});

describe('supportsConstructableStylesheets()', () => {
  it('should return true if CSSStyleSheet constructor exists', () => {
    // due to polyfill
    expect(supportsConstructableStylesheets()).toBe(true);
  });

  it('should return false if CSSStyleSheet constructor does not exist', () => {
    const globalCSSStyleSheet = global.CSSStyleSheet;
    delete global.CSSStyleSheet;
    expect(supportsConstructableStylesheets()).toBe(false);
    global.CSSStyleSheet = globalCSSStyleSheet;
  });
});

describe('buildSlottedStyles()', () => {
  it('should return @global styles object with node selector and important styles', () => {
    const el = document.createElement('p-button');
    expect(buildSlottedStyles(el, { div: { marginLeft: 5 } })).toStrictEqual({
      '@global': { 'p-button': { div: { marginLeft: '5 !important' } } },
    });
  });
});

describe('buildResponsiveHostStyles()', () => {
  describe('for simple getStyles', () => {
    const getStyles = (val: number): JssStyle => ({ width: 100 * val });

    it('should return flat jss for simple type', () => {
      expect(buildResponsiveHostStyles(6, getStyles)).toStrictEqual({ ':host': { width: 600 } });
    });

    it('should return nested jss for responsive type', () => {
      expect(buildResponsiveHostStyles({ base: 6, xs: 3, s: 4, m: 5, l: 6, xl: 7 }, getStyles)).toStrictEqual({
        ':host': {
          width: 600,
          '@media (min-width: 480px)': { width: 300 },
          '@media (min-width: 760px)': { width: 400 },
          '@media (min-width: 1000px)': { width: 500 },
          '@media (min-width: 1300px)': { width: 600 },
          '@media (min-width: 1760px)': { width: 700 },
        },
      });
    });
  });

  describe('for complex getStyles', () => {
    const getStyles = (val: number): JssStyle => ({ width: 100 * val, display: 'block' });

    it('should return flat jss for simple type', () => {
      expect(buildResponsiveHostStyles(6, getStyles)).toStrictEqual({ ':host': { width: 600, display: 'block' } });
    });

    it('should return nested jss for responsive type', () => {
      expect(buildResponsiveHostStyles({ base: 6, xs: 3, s: 4, m: 5, l: 6, xl: 7 }, getStyles)).toStrictEqual({
        ':host': {
          width: 600,
          display: 'block',
          '@media (min-width: 480px)': { width: 300, display: 'block' },
          '@media (min-width: 760px)': { width: 400, display: 'block' },
          '@media (min-width: 1000px)': { width: 500, display: 'block' },
          '@media (min-width: 1300px)': { width: 600, display: 'block' },
          '@media (min-width: 1760px)': { width: 700, display: 'block' },
        },
      });
    });
  });
});

describe('buildResponsiveStyles()', () => {
  describe('for simple getStyles', () => {
    const getStyles = (val: number): JssStyle => ({ width: 100 * val });

    it('should return flat jss for simple type', () => {
      expect(buildResponsiveStyles(6, getStyles)).toStrictEqual({ width: 600 });
    });

    it('should return nested jss for responsive type', () => {
      expect(buildResponsiveStyles({ base: 6, xs: 3, s: 4, m: 5, l: 6, xl: 7 }, getStyles)).toStrictEqual({
        width: 600,
        '@media (min-width: 480px)': { width: 300 },
        '@media (min-width: 760px)': { width: 400 },
        '@media (min-width: 1000px)': { width: 500 },
        '@media (min-width: 1300px)': { width: 600 },
        '@media (min-width: 1760px)': { width: 700 },
      });
    });
  });

  describe('for complex getStyles', () => {
    const getStyles = (val: number): JssStyle => ({ width: 100 * val, display: 'block' });

    it('should return flat jss for simple type', () => {
      expect(buildResponsiveStyles(6, getStyles)).toStrictEqual({ width: 600, display: 'block' });
    });

    it('should return nested jss for responsive type', () => {
      expect(buildResponsiveStyles({ base: 6, xs: 3, s: 4, m: 5, l: 6, xl: 7 }, getStyles)).toStrictEqual({
        width: 600,
        display: 'block',
        '@media (min-width: 480px)': { width: 300, display: 'block' },
        '@media (min-width: 760px)': { width: 400, display: 'block' },
        '@media (min-width: 1000px)': { width: 500, display: 'block' },
        '@media (min-width: 1300px)': { width: 600, display: 'block' },
        '@media (min-width: 1760px)': { width: 700, display: 'block' },
      });
    });
  });
});

describe('isObject()', () => {
  it('should return true for object', () => {
    expect(isObject({})).toBe(true);
    expect(isObject({ foo: 'bar' })).toBe(true);
  });

  it('should return false for other values', () => {
    expect(isObject('someString' as any)).toBe(false);
    expect(isObject(true as any)).toBe(false);
    expect(isObject([] as any)).toBe(false);
    expect(isObject(5 as any)).toBe(false);
  });
});

describe('mergeDeep()', () => {
  const data: { input: object[]; result: object }[] = [
    {
      input: [{}, { foo: 'bar' }],
      result: { foo: 'bar' },
    },
    {
      input: [{ foo: 'bar' }, {}, {}],
      result: { foo: 'bar' },
    },
    {
      input: [{ foo: 'bar' }, { xy: 1 }, { someVal: true }],
      result: { foo: 'bar', xy: 1, someVal: true },
    },
    {
      input: [{ foo: { key1: 'bar' } }, { foo: { key2: 'ok', key3: 'yea' }, some: 'thing' }],
      result: { foo: { key1: 'bar', key2: 'ok', key3: 'yea' }, some: 'thing' },
    },
    {
      input: [{ foo: { key1: 'bar' } }, { foo: { key1: 'ok' } }],
      result: { foo: { key1: 'ok' } },
    },
  ];
  it.each(
    data.map(({ input, result }) => [
      input.map((x) => JSON.stringify(x)).join(', '), // for test description
      JSON.stringify(result), // for test description
      input,
      result,
    ])
  )(`should be called with '%s' and return '%s'`, (_, __, input: object[], result: object) => {
    expect(mergeDeep(...input)).toStrictEqual(result);
  });
});

describe('attachComponentCss()', () => {
  beforeEach(() => {
    componentCssMap.clear();
  });

  it('should call getCachedComponentCss() with infinite parameters to retrieve cached css', () => {
    const host = document.createElement('p-some-component');
    host.attachShadow({ mode: 'open' });
    const spy = jest.spyOn(jssUtils, 'getCachedComponentCss').mockImplementation(() => '');

    attachComponentCss(host, (x: boolean) => 'some css', true);

    expect(spy).toHaveBeenCalledWith(host, expect.anything(), true);

    attachComponentCss(host, (x: boolean, y: string, z: number) => 'some css', false, '', 1);

    expect(spy).toHaveBeenCalledWith(host, expect.anything(), false, '', 1);
  });

  describe('with CSSStyleSheet support', () => {
    it('should create CSSStyleSheet and apply it to shadowRoot', () => {
      const div = document.createElement('p-some-component');
      div.attachShadow({ mode: 'open' });

      expect(div.shadowRoot.adoptedStyleSheets.length).toBe(0);

      attachComponentCss(div, () => ':host { display: "block" }');
      expect(div.shadowRoot.adoptedStyleSheets.length).toBe(1);
    });
  });

  describe('without CSSStyleSheet support', () => {
    it('should create style node and prepend it in shadowRoot', () => {
      const spy = jest.spyOn(jssUtils, 'supportsConstructableStylesheets').mockImplementation(() => false);

      const div = document.createElement('p-some-component');
      div.attachShadow({ mode: 'open' });
      expect(div.shadowRoot.querySelector('style')).toBeNull();

      const css = ':host { display: "block" }';
      attachComponentCss(div, () => css);
      expect(div.shadowRoot.querySelector('style').innerHTML).toBe(css);

      spy.mockRestore();
    });
  });
});

describe('getCachedComponentCss()', () => {
  beforeEach(() => {
    componentCssMap.clear();
  });

  it('should return css provided by css function', () => {
    const host = document.createElement('p-some-element');
    const getComponentCss = () => 'some css';

    expect(getCachedComponentCss(host, getComponentCss)).toBe('some css');
  });

  it('should call passed css function with infinite passed arguments', () => {
    const host = document.createElement('p-some-element');
    const getComponentCss1 = (a: number, b: boolean, c: string) => `some css ${a} ${b} ${c}`;

    expect(getCachedComponentCss(host, getComponentCss1, 1, true, 'some string')).toBe('some css 1 true some string');

    const getComponentCss2 = (d: { someProp: string }) => `some css ${d.someProp}`;

    expect(getCachedComponentCss(host, getComponentCss2, { someProp: 'some-object-value' })).toBe(
      'some css some-object-value'
    );
  });

  it('should keep CSS Cache clean and handle multiple types of infinite passed parameters', () => {
    const host1 = document.createElement('p-some-element');
    const host2 = document.createElement('my-prefix-p-some-element');
    const host3 = document.createElement('p-another-element');
    const getComponentCss1 = (a?: number, b?: boolean, c?: string, d?: { someProp: string }) => 'some css';

    getCachedComponentCss(host1, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host1, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host1, getComponentCss1);
    getCachedComponentCss(host1, getComponentCss1);

    getCachedComponentCss(host2, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host2, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host2, getComponentCss1);
    getCachedComponentCss(host2, getComponentCss1);

    getCachedComponentCss(host3, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host3, getComponentCss1, 1, true, 'some string', { someProp: 'some value' });
    getCachedComponentCss(host3, getComponentCss1);
    getCachedComponentCss(host3, getComponentCss1);

    expect(componentCssMap).toMatchSnapshot();
  });

  it('should call provided css function only once when it was already called before for the same host type', () => {
    const host1 = document.createElement('p-some-element');
    const host2 = document.createElement('p-some-element');
    const host3 = document.createElement('p-some-element');
    const getComponentCss = jest.fn();

    getCachedComponentCss(host1, getComponentCss, 'some-param');

    expect(getComponentCss).toHaveBeenCalledTimes(1);

    getCachedComponentCss(host2, getComponentCss, 'some-param');

    expect(getComponentCss).toHaveBeenCalledTimes(1);

    getCachedComponentCss(host3, getComponentCss, 'another-param');

    expect(getComponentCss).toHaveBeenCalledTimes(2);
  });

  it('should not call provided css function again for prefixed version of host type', () => {
    const host = document.createElement('p-some-element');
    const hostPrefixed = document.createElement('my-prefix-p-some-element');
    const getComponentCss = jest.fn();

    getCachedComponentCss(host, getComponentCss);

    expect(getComponentCss).toHaveBeenCalledTimes(1);

    getCachedComponentCss(hostPrefixed, getComponentCss);

    expect(getComponentCss).toHaveBeenCalledTimes(1);
  });
});
