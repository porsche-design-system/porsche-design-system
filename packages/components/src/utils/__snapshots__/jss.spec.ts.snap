// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCachedComponentCss() should keep CSS Cache clean and handle multiple types of infinite passed parameters 1`] = `
Map {
  "p-some-element" => Map {
    "1|true|some string|{"someProp":"some value"}" => "some css",
    "" => "some css",
  },
  "p-another-element" => Map {
    "1|true|some string|{"someProp":"some value"}" => "some css",
    "" => "some css",
  },
}
`;
